extensions/amp-fresh/0.1/amp-fresh-manager.js

```diff
@@ -14,20 +14,105 @@
  * limitations under the License.
  */
 
-import {fromClass} from '../../../src/service';
+import {addParamToUrl} from '../../../src/url';
+import {fromClassForDoc} from '../../../src/service';
+import {isExperimentOn} from '../../../src/experiments';
+import {whenDocumentReady} from '../../../src/document-ready';
+import {user} from '../../../src/log';
+import {xhrFor} from '../../../src/xhr';
 
 
+/** @const */
+const TAG = 'amp-fresh';
+
 /**
  * Manages registered amp-fresh components and does the xhr request
  * and update.
  */
-export class AmpFreshManager {}
+export class AmpFreshManager {
+
+  /**
+   * @param {!./ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+
+    /** @type {!./ampdoc-impl.AmpDoc} */
+    this.ampdoc = ampdoc;
+
+    if (!isExperimentOn(this.win, TAG)) {
+      return;
+    }
+
+    /** @private @const {!Object<string, !./amp-fresh.AmpFresh>} */
+    this.ampFreshInstances_ = Object.create(null);
+
+    /** @private {?Document} */
+    this.docPromise_ = null;
+
+    this.fetchDocument_().then(doc => {
+      this.update_(doc);
+    });
+  }
+
+  /**
+   * Registers an amp-fresh component instance to be managed by this windows
+   * amp-fresh-manager instance.
+   * @param {string} id
+   * @param {!./amp-fresh.AmpFresh} ampFreshIntance
+   */
+  register(id, ampFreshInstance) {
+    user().assert(!this.ampFreshInstances_[id],
+        `duplicate amp-fresh id used: ${id}`);
+    this.ampFreshInstances_[id] = ampFreshInstance;
+  }
+
+  /**
+   * @return {!Promise<!Document>}
+   */
+  fetchDocument_() {
+    if (this.docPromise_) {
+      return this.docPromise_;
+    }
+    const url = addParamToUrl(this.ampdoc.win.location.href,
+        'amp-fresh', '1');
+    return this.docPromise_ = Promise.all([
+      xhrFor(this.ampdoc.win).fetchDocument(url),
+      this.ampdoc.whenReady(),
+    ]).then(args => args[0])
+    .catch(() => {
+      this.onFetchDocumentFailure_();
+    });
+  }
+
+  /**
+   * @param {!Document} docFromServer
+   */
+  update_(docFromServer) {
+    const elements = docFromServer.getElementsByTagName('amp-fresh');
+    for (let i = 0; i < elements.length; i++) {
+      const ampFreshFromServer = elements[i];
+      const id = ampFreshFromServer.getAttribute('id');
+      // the amp-fresh instance live in client DOM.
+      const ampFresh = this.ampFreshInstances_[id];
+      if (ampFresh) {
+        ampFresh.update(ampFreshFromServer);
+      }
+    }
+  }
+
+  onFetchDocumentFailure_() {
+    whenDocumentReady(this.win.document).then(() => {
+      Object.keys(this.ampFreshInstances_).forEach(id => {
```

Comment:
We'd need to polyfill `Object.values`, and it'd always be slower than `keys` since we'd loop all keys just to get values, then loop the values. I'd say let's just stick with keys in the codebase.